typing/typecore.ml

```diff
@@ -6411,28 +6482,48 @@ and type_expect_
               exp_attributes = sexp.pexp_attributes;
               exp_env = env }
         end
-  | Pexp_setinstvar (lab, snewval) -> begin
-      let (path, mut, cl_num, ty) =
-        Env.lookup_instance_variable ~loc lab.txt env
-      in
-      match mut with
-      | Mutable ->
-          let newval =
-            type_expect env mode_legacy snewval
-              (mk_expected (instance ty))
-          in
-          let (path_self, _) =
-            Env.find_value_by_name (Longident.Lident ("self-" ^ cl_num)) env
-          in
-          rue {
-            exp_desc = Texp_setinstvar(path_self, path, lab, newval);
-            exp_loc = loc; exp_extra = [];
-            exp_type = instance Predef.type_unit;
-            exp_attributes = sexp.pexp_attributes;
-            exp_env = env }
-      | _ ->
-          raise(Error(loc, env, Instance_variable_not_mutable lab.txt))
-    end
+  | Pexp_setvar (lab, snewval) ->
+      let desc =
+        match Env.lookup_settable_variable ~loc lab.txt env with
+        | Instance_variable (path,Mutable,cl_num,ty) ->
+            let newval =
+              type_expect env
+                (mode_default (Mode.Value.max_with (Comonadic Areality)
+                  Mode.Regionality.global))
+                snewval (mk_expected (instance ty))
+            in
+            let (path_self, _) =
+              Env.find_value_by_name (Longident.Lident ("self-" ^ cl_num)) env
+            in
+            Texp_setinstvar(path_self, path, lab, newval)
+        | Instance_variable (_,Immutable,_,_) ->
+            raise(Error(loc, env, Instance_variable_not_mutable lab.txt))
+        | Mutable_variable (id,mode,ty) ->
+            let newval =
+              type_expect env (mode_default mode)
+                snewval (mk_expected (instance ty))
+            in
+            let lid = {txt = id; loc} in
+            let sort =
+              match
+                Ctype.type_sort ~why:Jkind.History.Mutable_var_assignment
+                  ~fixed:false env newval.exp_type
+              with
+              | Ok sort -> sort
+              | Error _ ->
+                (* unreachable since [x <- e] was already type-checked,
+                   so [e] is representable *)
```

Comment:
I think this isn't quite right. We're in the middle of checking `x <- e`, here.  It's unreachable because we checked `ty` is representable when we created the binding for the mutable variable.

Which, actually, gives me an idea: could we just remember the sort of this variable in the environment rather than recomputing it here?